diff -Nrup a/Makefile.am b/Makefile.am
--- a/Makefile.am	2016-07-22 16:46:33.000000000 -0400
+++ b/Makefile.am	2016-07-22 16:49:00.260462649 -0400
@@ -37,7 +37,7 @@ endif
 SUBDIRS = tests $(TESTS_M32) $(TESTS_MX32)
 
 bin_PROGRAMS = strace
-man_MANS = strace.1
+man_MANS = strace.1 strace-log-merge.1
 bin_SCRIPTS = strace-graph strace-log-merge
 
 OS		= linux
diff -Nrup a/Makefile.in b/Makefile.in
--- a/Makefile.in	2016-07-22 16:46:33.000000000 -0400
+++ b/Makefile.in	2016-07-22 16:49:00.267462671 -0400
@@ -610,7 +610,7 @@ top_srcdir = @top_srcdir@
 @HAVE_M32_RUNTIME_TRUE@TESTS_M32 = tests-m32
 @HAVE_MX32_RUNTIME_TRUE@TESTS_MX32 = tests-mx32
 SUBDIRS = tests $(TESTS_M32) $(TESTS_MX32)
-man_MANS = strace.1
+man_MANS = strace.1 strace-log-merge.1
 bin_SCRIPTS = strace-graph strace-log-merge
 OS = linux
 # ARCH is `i386', `m68k', `sparc', etc.
diff -Nrup a/strace-log-merge.1 b/strace-log-merge.1
--- a/strace-log-merge.1	1969-12-31 19:00:00.000000000 -0500
+++ b/strace-log-merge.1	2016-07-22 16:49:00.268462674 -0400
@@ -0,0 +1,18 @@
+.TH strace-log-merge 1 "15 April 2013" ".1" "strace-log-merge"
+.SH NAME
+strace-log-merge
+.SH SYNOPSIS
+strace-log-merge STRACE_LOG
+.SH DESCRIPTION
+strace-log-merge finds all STRACE_LOG.PID files, adds PID prefix to every line,
+the combines and sorts them, and prints the result to standard output.
+
+It is assumed that STRACE_LOGs were produced by strace with the -tt[t]
+option which prints timestamps (otherwise worting won't do any good).
+
+.SH OPTIONS
+--help  Show help
+
+.SH SEE ALSO
+.SH BUGS
+
